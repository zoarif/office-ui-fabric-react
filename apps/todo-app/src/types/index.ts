export * from './ITodo';
export * from './ITodoTask';
export * from './ITodoTabs';
export * from './IDataProvider';