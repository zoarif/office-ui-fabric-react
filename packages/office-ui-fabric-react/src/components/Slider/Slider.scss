@import '../../common/common';

.root {
  user-select: none;
}

.titleLabel {
  padding: 0;
 }

.line {
  display: flex;
  position: relative;
}

.line .lineContainer {
  border-radius: 4px;
  box-sizing: border-box;
}

.container {
  display: flex;
  flex-wrap: nowrap;
  align-items: center;
}

.valueLabel {
  flex-shrink: 1;
  width: 30px;
  line-height: 1;
}

.slideBox {
  @include focus-border();
  background: transparent;
  border: none;
  flex-grow: 1;
  line-height: 28px;
}

.slideBox .thumb {
  border: 2px solid $ms-color-neutralSecondary;
  box-sizing: border-box;
  background: $ms-color-white;
  display: block;
  width: 16px;
  height: 16px;
  position: absolute;
  border-radius: 10px;
}

.rootIsHorizontal {
  .line {
    width: 100%;
  }

  .line .lineContainer {
    height: 4px;
  }

  .slideBox {
    height: 28px;
    padding: 0 8px; // Make room for thumb at ends of line
  }

  .slideBox .thumb {
    top: -6px;
    @include transform-translateX(-50%);
  }

  .valueLabel {
    margin: 0 8px;
    white-space: nowrap;
    width: 40px;
  }
}

.rootIsVertical {
  margin-right: 8px;

  .line {
    height: 100%;
    width: 4px;
    margin: 0 auto;
    flex-direction: column-reverse;
  }

  .line .lineContainer {
    width: 4px;
    height: 100%;
  }

  .slideBox {
    height: 100%;
    width: 28px;
    padding: 8px 0; // Make room for thumb at bottom of line
  }

  .slideBox .thumb {
    left: -6px;
    margin: 0 auto;
    transform: translateY(8px);
  }

  .container {
    flex-direction: column;
    height: 100%;
    text-align: center;
    margin: 8px 0;
  }

  .valueLabel {
    margin: 0 auto;
    white-space: nowrap;
    width: 40px;
  }
}

.activeSection {
  background: $ms-color-neutralSecondary;

  @include high-contrast {
    background-color: WindowText;
  }
}

.inactiveSection {
  background: $ms-color-neutralTertiaryAlt;

  @include high-contrast {
    border: 1px solid WindowText;
  }
}

.showTransitions {
  .thumb {
    transition: left $ms-animation-duration-3 $ms-animation-ease-1;
  }

  .activeSection,
  .inactiveSection {
    transition: width $ms-animation-duration-3 $ms-animation-ease-1;
  }
}

.rootIsEnabled {

  .slideBox:hover,
  .slideBox:active {

    .thumb {
      border: 2px solid $ms-color-themePrimary;

      @include high-contrast {
        border-color: Highlight;
      }
    }

    .activeSection {
      background-color: $ms-color-themePrimary;

      @include high-contrast {
        background-color: Highlight;
      }
    }

    .inactiveSection {
      background-color: $ms-color-themeLight;

      @include high-contrast {
        border-color: Highlight;
      }
    }
  }

  .slideBox:active {
    .thumb {
      border: 2px solid $ms-color-themeDarkAlt;
    }
    .activeSection {
      background-color: $ms-color-themeDarkAlt;
    }
  }
}

.rootIsDisabled {
  .thumb {
    border-color: $ms-color-neutralTertiaryAlt;

    @include high-contrast {
      border-color: GrayText;
    }
  }

  .activeSection,
  .inactiveSection {
    @include high-contrast {
      background-color: GrayText;
      border-color: GrayText;
    }
  }

  .activeSection {
    background: $ms-color-neutralTertiaryAlt;
  }
  .inactiveSection {
    background: $ms-color-neutralLight;
  }
}
