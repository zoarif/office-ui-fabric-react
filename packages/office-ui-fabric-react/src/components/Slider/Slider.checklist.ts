import { ChecklistStatus } from '../../demo/ComponentStatus/ComponentStatus.types';

export const SliderStatus = {
  keyboardAccessibilitySupport: ChecklistStatus.unknown,
  markupSupport: ChecklistStatus.unknown,
  highContrastSupport: ChecklistStatus.pass,
  rtlSupport: ChecklistStatus.pass,
  testCoverage: ChecklistStatus.fair
};