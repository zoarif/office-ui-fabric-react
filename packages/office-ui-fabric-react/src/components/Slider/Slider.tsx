import * as React from 'react';
import {
  BaseComponent,
  KeyCodes,
  css,
  getId,
  getRTL,
  getRTLSafeKeyCode,
  createRef
} from '../../Utilities';
import { ISliderProps, ISlider } from './Slider.types';
import { Label } from '../../Label';
import * as stylesImport from './Slider.scss';
const styles: any = stylesImport;

export interface ISliderState {
  value?: number;
  renderedValue?: number;
}

export enum ValuePosition {
  Previous = 0,
  Next = 1
}

export class Slider extends BaseComponent<ISliderProps, ISliderState> implements ISlider {
  public static defaultProps: {} = {
    step: 1,
    min: 0,
    max: 10,
    showValue: true,
    disabled: false,
    vertical: false,
    buttonProps: {}
  };

  private _sliderLine = createRef<HTMLDivElement>();
  private _thumb = createRef<HTMLSpanElement>();
  private _id: string;

  constructor(props: ISliderProps) {
    super(props);

    this._warnMutuallyExclusive({
      'value': 'defaultValue'
    });

    this._id = getId('Slider');

    const value = props.value || props.defaultValue || props.min;

    this.state = {
      value: value,
      renderedValue: value
    };
  }

  /**
   * Invoked when a component is receiving new props. This method is not called for the initial render.
   */
  public componentWillReceiveProps(newProps: ISliderProps): void {

    if (newProps.value !== undefined) {
      const value = Math.max(newProps.min as number, Math.min(newProps.max as number, newProps.value));

      this.setState({
        value: value,
        renderedValue: value
      });
    }
  }

  public render(): React.ReactElement<{}> {
    const {
      ariaLabel,
      className,
      disabled,
      label,
      max,
      min,
      showValue,
      buttonProps,
      vertical
    } = this.props;
    const { value, renderedValue } = this.state;
    const thumbOffsetPercent: number = (renderedValue! - min!) / (max! - min!) * 100;
    const lengthString = vertical ? 'height' : 'width';
    const onMouseDownProp: {} = disabled ? {} : { onMouseDown: this._onMouseDownOrTouchStart };
    const onTouchStartProp: {} = disabled ? {} : { onTouchStart: this._onMouseDownOrTouchStart };
    const onKeyDownProp: {} = disabled ? {} : { onKeyDown: this._onKeyDown };

    return (
      <div
        className={ css('ms-Slider', styles.root, className, {
          ['ms-Slider-enabled ' + styles.rootIsEnabled]: !disabled,
          ['ms-Slider-disabled ' + styles.rootIsDisabled]: disabled,
          ['ms-Slider-row ' + styles.rootIsHorizontal]: !vertical,
          ['ms-Slider-column ' + styles.rootIsVertical]: vertical
        }) }
      >
        { label && (
          <Label className={ styles.titleLabel } { ...ariaLabel ? {} : { 'htmlFor': this._id } }>
            { label }
          </Label>
        ) }
        <div className={ css('ms-Slider-container', styles.container) }>
          <button
            aria-valuenow={ value }
            aria-valuemin={ min }
            aria-valuemax={ max }
            aria-valuetext={ this._getAriaValueText(value) }
            aria-label={ ariaLabel || label }
            { ...onMouseDownProp }
            { ...onTouchStartProp }
            { ...onKeyDownProp }
            { ...buttonProps }
            className={ css(
              'ms-Slider-slideBox',
              styles.slideBox,
              buttonProps!.className,
              !!showValue && 'ms-Slider-showValue',
              (renderedValue === value) && ('ms-Slider-showTransitions ' + styles.showTransitions)
            ) }
            id={ this._id }
            disabled={ disabled }
            type='button'
            role='slider'
          >
            <div
              ref={ this._sliderLine }
              className={ css('ms-Slider-line', styles.line) }
            >
              <span
                ref={ this._thumb }
                className={ css('ms-Slider-thumb', styles.thumb) }
                style={ this._getThumbStyle(vertical, thumbOffsetPercent) }
              />
              <span
                className={ css('ms-Slider-active', styles.lineContainer, styles.activeSection) }
                style={ { [lengthString]: thumbOffsetPercent + '%' } }
              />
              <span
                className={ css('ms-Slider-inactive', styles.lineContainer, styles.inactiveSection) }
                style={ { [lengthString]: (100 - thumbOffsetPercent) + '%' } }
              />
            </div>
          </button>
          { showValue && <Label className={ css('ms-Slider-value', styles.valueLabel) }>{ value }</Label> }
        </div>
      </div>
    ) as React.ReactElement<{}>;
  }
  public focus(): void {
    if (this._thumb.current) {
      this._thumb.current.focus();
    }
  }

  public get value(): number | undefined {
    return this.state.value;
  }

  private _getAriaValueText = (value: number | undefined): string | undefined => {
    if (this.props.ariaValueText && value !== undefined) {
      return this.props.ariaValueText(value);
    }
  }

  private _getThumbStyle(vertical: boolean | undefined, thumbOffsetPercent: number): any {
    const direction: string = vertical ? 'bottom' : (getRTL() ? 'right' : 'left');
    return {
      [direction]: thumbOffsetPercent + '%'
    };
  }

  private _onMouseDownOrTouchStart = (event: MouseEvent | TouchEvent): void => {
    if (event.type === 'mousedown') {
      this._events.on(window, 'mousemove', this._onMouseMoveOrTouchMove, true);
      this._events.on(window, 'mouseup', this._onMouseUpOrTouchEnd, true);
    } else if (event.type === 'touchstart') {
      this._events.on(window, 'touchmove', this._onMouseMoveOrTouchMove, true);
      this._events.on(window, 'touchend', this._onMouseUpOrTouchEnd, true);
    }
    this._onMouseMoveOrTouchMove(event, true);
  }

  private _onMouseMoveOrTouchMove = (event: MouseEvent | TouchEvent, suppressEventCancelation?: boolean): void => {
    if (!this._sliderLine.current) {
      return;
    }

    const { max, min, step } = this.props;
    const steps: number = (max! - min!) / step!;
    const sliderPositionRect: ClientRect = this._sliderLine.current.getBoundingClientRect();
    const sliderLength: number = !this.props.vertical ? sliderPositionRect.width : sliderPositionRect.height;
    const stepLength: number = sliderLength / steps;
    let currentSteps: number | undefined;
    let distance: number | undefined;

    if (!this.props.vertical) {
      const left: number | undefined = this._getPosition(event, this.props.vertical);
      distance = getRTL() ? sliderPositionRect.right - left! : left! - sliderPositionRect.left;
      currentSteps = distance / stepLength;
    } else {
      const bottom: number | undefined = this._getPosition(event, this.props.vertical);
      distance = sliderPositionRect.bottom - bottom!;
      currentSteps = distance / stepLength;
    }

    let currentValue: number | undefined;
    let renderedValue: number | undefined;

    // The value shouldn't be bigger than max or be smaller than min.
    if (currentSteps! > Math.floor(steps)) {
      renderedValue = currentValue = max as number;
    } else if (currentSteps! < 0) {
      renderedValue = currentValue = min as number;
    } else {
      renderedValue = min! + step! * currentSteps!;
      currentValue = min! + step! * Math.round(currentSteps!);
    }

    this._updateValue(currentValue, renderedValue);

    if (!suppressEventCancelation) {
      event.preventDefault();
      event.stopPropagation();
    }
  }

  private _getPosition(event: MouseEvent | TouchEvent, vertical: boolean | undefined): number | undefined {
    let currentPosition: number | undefined;
    switch (event.type) {
      case 'mousedown':
      case 'mousemove':
        currentPosition = !vertical ? (event as MouseEvent).clientX : (event as MouseEvent).clientY;
        break;
      case 'touchstart':
      case 'touchmove':
        currentPosition = !vertical ? (event as TouchEvent).touches[0].clientX : (event as TouchEvent).touches[0].clientY;
        break;
    }
    return currentPosition;
  }
  private _updateValue(value: number, renderedValue: number): void {
    const interval: number = 1.0 / this.props.step!;
    // Make sure value has correct number of decimal places based on steps without JS's floating point issues
    const roundedValue: number = Math.round(value * interval) / interval;

    const valueChanged = roundedValue !== this.state.value;

    this.setState({
      value: roundedValue,
      renderedValue
    }, () => {
      if (valueChanged && this.props.onChange) {
        this.props.onChange(this.state.value as number);
      }
    });
  }

  private _onMouseUpOrTouchEnd = (): void => {
    // Synchronize the renderedValue to the actual value.
    this.setState({
      renderedValue: this.state.value
    });

    this._events.off();
  }

  private _onKeyDown = (event: KeyboardEvent): void => {
    let value: number | undefined = this.state.value;
    const { max, min, step } = this.props;

    let diff: number | undefined = 0;

    switch (event.which) {
      case getRTLSafeKeyCode(KeyCodes.left):
      case KeyCodes.down:
        diff = -(step as number);
        break;
      case getRTLSafeKeyCode(KeyCodes.right):
      case KeyCodes.up:
        diff = step;
        break;

      case KeyCodes.home:
        value = min;
        break;

      case KeyCodes.end:
        value = max;
        break;

      default:
        return;
    }

    const newValue: number = Math.min(max as number, Math.max(min as number, value! + diff!));

    this._updateValue(newValue, newValue);

    event.preventDefault();
    event.stopPropagation();
  }
}
