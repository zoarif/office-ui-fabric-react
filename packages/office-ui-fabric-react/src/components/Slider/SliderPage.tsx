import * as React from 'react';
import {
  ExampleCard,
  IComponentDemoPageProps,
  ComponentPage,
  PageMarkdown,
  PropertiesTableSet
} from '@uifabric/example-app-base';
import { SliderBasicExample } from './examples/Slider.Basic.Example';
import { ComponentStatus } from '../../demo/ComponentStatus/ComponentStatus';
import { SliderStatus } from './Slider.checklist';

const SliderBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/Slider/examples/Slider.Basic.Example.tsx') as string;

export class SliderPage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title={ 'Slider' }
        componentName='SliderExample'
        componentUrl='https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/Slider'
        exampleCards={
          <ExampleCard title={ 'Slider' } code={ SliderBasicExampleCode }>
            <SliderBasicExample />
          </ExampleCard>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={ [
              require<string>('!raw-loader!office-ui-fabric-react/src/components/Slider/Slider.types.ts')
            ] }
          />
        }
        overview={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderOverview.md') }
          </PageMarkdown>
        }
        bestPractices={
          <div />
        }
        dos={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderDos.md') }
          </PageMarkdown>
        }
        donts={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/Slider/docs/SliderDonts.md') }
          </PageMarkdown>
        }
        isHeaderVisible={ this.props.isHeaderVisible }
        componentStatus={
          <ComponentStatus
            { ...SliderStatus }
          />
        }
      />
    );
  }
}
