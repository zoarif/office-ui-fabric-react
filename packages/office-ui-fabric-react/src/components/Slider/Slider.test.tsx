import * as React from 'react';

import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';

import { Slider } from './Slider';
import { ISlider } from './Slider.types';

describe('Slider', () => {

  it('renders Slider correctly', () => {
    const component = renderer.create(<Slider />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a slider', () => {
    const component = ReactTestUtils.renderIntoDocument(
      <Slider label='slider' />
    );
    const renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance) as Element;
    const labelElement = renderedDOM.querySelector('.ms-Label') as HTMLElement;

    expect(labelElement.textContent).toEqual('slider');
  });

  it('can slide to default min/max and execute onChange', () => {
    let changedValue;
    const onChange = (val: any) => {
      changedValue = val;
    };
    const component = ReactTestUtils.renderIntoDocument<React.ReactInstance>(
      <Slider
        onChange={ onChange }
      />
    );

    const renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance) as Element;
    const sliderLine = renderedDOM.querySelector('.ms-Slider-line') as HTMLElement;
    const sliderThumb = renderedDOM.querySelector('.ms-Slider-slideBox') as HTMLElement;

    sliderLine.getBoundingClientRect = () => ({
      left: 0,
      top: 0,
      right: 100,
      bottom: 40,
      width: 100,
      height: 40
    });

    ReactTestUtils.Simulate.mouseDown(sliderThumb, {
      type: 'mousedown',
      clientX: 100,
      clientY: 0
    });

    // Default max is 10.
    expect(changedValue).toEqual(10);

    ReactTestUtils.Simulate.mouseDown(sliderThumb, {
      type: 'mousedown',
      clientX: 0,
      clientY: 0
    });

    // Default min is 0.
    expect(changedValue).toEqual(0);
  });

  it('has type=button on all buttons', () => {
    const component = ReactTestUtils.renderIntoDocument<React.ReactInstance>(
      <Slider />
    );

    const renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance) as Element;
    const allButtons = renderedDOM.querySelectorAll('button');

    for (let i = 0; i < allButtons.length; i++) {
      const button = allButtons[i];

      expect(button.getAttribute('type')).toEqual('button');
    }
  });

  it('can read the current value', () => {
    let slider: ISlider | null;

    ReactTestUtils.renderIntoDocument(
      // tslint:disable-next-line:jsx-no-lambda
      <Slider label='slider' defaultValue={ 12 } min={ 0 } max={ 100 } componentRef={ s => slider = s } />
    );
    expect(slider!.value).toEqual(12);
  });

  it('renders correct aria-valuetext', () => {
    let component = ReactTestUtils.renderIntoDocument(
      <Slider />
    );
    let renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance) as Element;
    let button = renderedDOM.querySelector('.ms-Slider-slideBox') as HTMLElement;
    let ariaValueText = button.getAttribute('aria-valuetext');

    expect(ariaValueText).toBeNull();

    const values = ['small', 'medium', 'large'];
    const selected = 1;
    const getTextValue = (value: number) => values[value];

    component = ReactTestUtils.renderIntoDocument(
      <Slider
        value={ selected }
        ariaValueText={ getTextValue }
      />
    );
    renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance) as Element;
    button = renderedDOM.querySelector('.ms-Slider-slideBox') as HTMLElement;
    ariaValueText = button.getAttribute('aria-valuetext');

    expect(ariaValueText).toEqual(values[selected]);
  });
});
