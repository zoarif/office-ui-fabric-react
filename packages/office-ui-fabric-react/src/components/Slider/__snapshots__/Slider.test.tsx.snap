// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders Slider correctly 1`] = `
<div
  className="ms-Slider ms-Slider-enabled undefined ms-Slider-row undefined"
>
  <div
    className="ms-Slider-container"
  >
    <button
      aria-label={undefined}
      aria-valuemax={10}
      aria-valuemin={0}
      aria-valuenow={0}
      aria-valuetext={undefined}
      className="ms-Slider-slideBox ms-Slider-showValue ms-Slider-showTransitions undefined"
      disabled={false}
      id="Slider0"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      role="slider"
      type="button"
    >
      <div
        className="ms-Slider-line"
      >
        <span
          className="ms-Slider-thumb"
          style={
            Object {
              "left": "0%",
            }
          }
        />
        <span
          className="ms-Slider-active"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <span
          className="ms-Slider-inactive"
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
    </button>
    <label
      className=
          ms-Label
          ms-Slider-value
          {
            box-shadow: none;
            box-sizing: border-box;
            color: #333333;
            display: block;
            margin-bottom: 0px;
            margin-left: 0px;
            margin-right: 0px;
            margin-top: 0px;
            overflow-wrap: break-word;
            padding-bottom: 5px;
            padding-left: 0;
            padding-right: 0;
            padding-top: 5px;
            word-wrap: break-word;
          }
    >
      0
    </label>
  </div>
</div>
`;
